File: python/app/pfm/streamline_pfms.py
"""Streamline PFMs — flow-field line tracing."""

from __future__ import annotations

import math
from typing import Callable, List

from app.pfm import (
    DrawingGeometry, PathFindingModule, PFMSetting, SettingType,
    make_plotting_resolution_setting, make_random_seed_setting,
)
from app.pfm.pfm_utils import flow_field, prepare_work_image, trace_streamline


def _trace_streamlines(work, fx, fy, pfm, progress) -> List[DrawingGeometry]:
    h, w = work.shape
    geoms: List[DrawingGeometry] = []
    count = int(pfm.get("line_count"))
    for i in range(count):
        if pfm.is_cancelled:
            break
        x0 = pfm._rng.uniform(0, w)
        y0 = pfm._rng.uniform(0, h)
        path = trace_streamline(
            fx, fy, x0, y0,
            int(pfm.get("max_length")),
            float(pfm.get("step_size")),
        )
        if len(path) >= 2:
            geoms.append(DrawingGeometry(path=path))
        if i % 20 == 0:
            progress(i / count, len(geoms), "Tracing streamlines...")
    return geoms


class _StreamlineBase(PathFindingModule):
    category = "Streamline"
    is_premium = True

    def _base_settings(self) -> List[PFMSetting]:
        return [
            make_plotting_resolution_setting(),
            make_random_seed_setting(),
            PFMSetting("line_count", "Line Count", SettingType.INTEGER, 120,
                       10, 2000, 50, category="Streamlines"),
            PFMSetting("max_length", "Max Length", SettingType.INTEGER, 200,
                       20, 2000, 10, category="Streamlines"),
            PFMSetting("step_size", "Step Size", SettingType.NUMBER, 1.5,
                       0.5, 5.0, 0.1, category="Streamlines"),
        ]


class StreamlinesEdgeFieldPFM(_StreamlineBase):
    @property
    def name(self) -> str:
        return "Streamlines Edge Field"

    def _define_settings(self) -> List[PFMSetting]:
        return self._base_settings()

    def _process(self, image, progress) -> List[DrawingGeometry]:
        work, w, h = prepare_work_image(image, self.get("plotting_resolution"))
        fx, fy = flow_field(work, smooth=7)
        return _trace_streamlines(work, -fy, fx, self, progress)


class StreamlinesFlowFieldPFM(_StreamlineBase):
    @property
    def name(self) -> str:
        return "Streamlines Flow Field"

    def _define_settings(self) -> List[PFMSetting]:
        return self._base_settings()

    def _process(self, image, progress) -> List[DrawingGeometry]:
        work, w, h = prepare_work_image(image, self.get("plotting_resolution"))
        fx, fy = flow_field(work, smooth=5)
        return _trace_streamlines(work, fx, fy, self, progress)


class StreamlinesSuperformulaPFM(_StreamlineBase):
    @property
    def name(self) -> str:
        return "Streamlines Superformula"

    def _define_settings(self) -> List[PFMSetting]:
        s = self._base_settings()
        s.append(PFMSetting("wave_amp", "Wave Amplitude", SettingType.NUMBER, 2.0,
                            0.0, 20.0, 0.5, category="Style"))
        return s

    def _process(self, image, progress) -> List[DrawingGeometry]:
        work, w, h = prepare_work_image(image, self.get("plotting_resolution"))
        fx, fy = flow_field(work, smooth=3)
        amp = float(self.get("wave_amp"))
        geoms: List[DrawingGeometry] = []
        count = int(self.get("line_count"))
        for i in range(count):
            if self.is_cancelled:
                break
            x0 = self._rng.uniform(0, w)
            y0 = self._rng.uniform(0, h)
            path = trace_streamline(fx, fy, x0, y0, int(self.get("max_length")),
                                    float(self.get("step_size")))
            if len(path) >= 2 and amp > 0:
                path = [
                    (x + math.sin(j * 0.2) * amp, y + math.cos(j * 0.15) * amp)
                    for j, (x, y) in enumerate(path)
                ]
            if len(path) >= 2:
                geoms.append(DrawingGeometry(path=path))
            if i % 20 == 0:
                progress(i / count, len(geoms), "Superformula streamlines...")
        return geoms


STREAMLINE_PFMS = [
    StreamlinesEdgeFieldPFM(),
    StreamlinesFlowFieldPFM(),
    StreamlinesSuperformulaPFM(),
]

File: cpp/pfm/streamline_pfms.cpp
#include "pfm/streamline_pfms.h"
#include <opencv2/imgproc.hpp>
#include <cmath>

static Path trace_streamline(const cv::Mat& fx, const cv::Mat& fy, float x, float y, int max_len, float step_size) {
    Path path;
    int w = fx.cols;
    int h = fx.rows;
    for (int i = 0; i < max_len; ++i) {
        int xi = std::clamp(int(x), 0, w - 1);
        int yi = std::clamp(int(y), 0, h - 1);
        path.push_back({x, y});
        
        float vx = fx.at<float>(yi, xi);
        float vy = fy.at<float>(yi, xi);
        float mag = std::hypot(vx, vy);
        if (mag < 1e-6f) break;
        
        x += (vx / mag) * step_size;
        y += (vy / mag) * step_size;
        if (x < 0 || x >= w || y < 0 || y >= h) break;
    }
    return path;
}

BaseStreamlinePFM::BaseStreamlinePFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> BaseStreamlinePFM::defineSettings() const {
    return {
        {"line_count", "Line Count", SettingType::Integer, 120, QVariant(), 10, 2000, 10, 2000, 50},
        {"max_length", "Max Length", SettingType::Integer, 200, QVariant(), 20, 2000, 20, 2000, 10},
        {"step_size", "Step Size", SettingType::Number, 1.5, QVariant(), 0.5, 5.0, 0.5, 5.0, 0.1}
    };
}

QVector<DrawingGeometry> StreamlinesEdgeFieldPFM::_process(const cv::Mat& image) {
    int count = m_settings["line_count"].toInt();
    int max_len = m_settings["max_length"].toInt();
    float step = m_settings["step_size"].toDouble();
    
    cv::Mat gx, gy;
    cv::Sobel(image, gx, CV_32F, 1, 0, 3);
    cv::Sobel(image, gy, CV_32F, 0, 1, 3);
    cv::GaussianBlur(gx, gx, cv::Size(7, 7), 0);
    cv::GaussianBlur(gy, gy, cv::Size(7, 7), 0);
    
    // For Edge field, flow is perpendicular to gradient
    cv::Mat fx = -gy;
    cv::Mat fy = gx;
    
    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < count; ++i) {
        if (isCancelled()) break;
        if (i % 20 == 0) emitProgress(float(i) / count, geoms.size(), "Edge Field Streamlines...");
        float x = randUniform(0, image.cols);
        float y = randUniform(0, image.rows);
        Path path = trace_streamline(fx, fy, x, y, max_len, step);
        if (path.size() >= 2) { DrawingGeometry dg; dg.path = path; geoms.push_back(dg); }
    }
    return geoms;
}

QVector<DrawingGeometry> StreamlinesFlowFieldPFM::_process(const cv::Mat& image) {
    int count = m_settings["line_count"].toInt();
    int max_len = m_settings["max_length"].toInt();
    float step = m_settings["step_size"].toDouble();
    
    cv::Mat gx, gy;
    cv::Sobel(image, gx, CV_32F, 1, 0, 3);
    cv::Sobel(image, gy, CV_32F, 0, 1, 3);
    cv::GaussianBlur(gx, gx, cv::Size(5, 5), 0);
    cv::GaussianBlur(gy, gy, cv::Size(5, 5), 0);
    
    // For Flow field, flow is along gradient
    cv::Mat fx = gx;
    cv::Mat fy = gy;
    
    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < count; ++i) {
        if (isCancelled()) break;
        if (i % 20 == 0) emitProgress(float(i) / count, geoms.size(), "Flow Field Streamlines...");
        float x = randUniform(0, image.cols);
        float y = randUniform(0, image.rows);
        Path path = trace_streamline(fx, fy, x, y, max_len, step);
        if (path.size() >= 2) { DrawingGeometry dg; dg.path = path; geoms.push_back(dg); }
    }
    return geoms;
}

QVector<PFMSetting> StreamlinesSuperformulaPFM::defineSettings() const {
    auto s = BaseStreamlinePFM::defineSettings();
    s.push_back({"wave_amp", "Wave Amplitude", SettingType::Number, 2.0, QVariant(), 0.0, 20.0, 0.0, 20.0, 0.5});
    return s;
}

QVector<DrawingGeometry> StreamlinesSuperformulaPFM::_process(const cv::Mat& image) {
    int count = m_settings["line_count"].toInt();
    int max_len = m_settings["max_length"].toInt();
    float step = m_settings["step_size"].toDouble();
    float amp = m_settings["wave_amp"].toDouble();
    
    cv::Mat gx, gy;
    cv::Sobel(image, gx, CV_32F, 1, 0, 3);
    cv::Sobel(image, gy, CV_32F, 0, 1, 3);
    cv::GaussianBlur(gx, gx, cv::Size(3, 3), 0);
    cv::GaussianBlur(gy, gy, cv::Size(3, 3), 0);
    
    cv::Mat fx = gx;
    cv::Mat fy = gy;
    
    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < count; ++i) {
        if (isCancelled()) break;
        if (i % 20 == 0) emitProgress(float(i) / count, geoms.size(), "Superformula Streamlines...");
        float x = randUniform(0, image.cols);
        float y = randUniform(0, image.rows);
        Path path = trace_streamline(fx, fy, x, y, max_len, step);
        if (path.size() >= 2 && amp > 0) {
            Path waved;
            for (size_t j = 0; j < path.size(); ++j) {
                float wx = path[j].first + std::sin(j * 0.2f) * amp;
                float wy = path[j].second + std::cos(j * 0.15f) * amp;
                waved.push_back({wx, wy});
            }
            path = waved;
        }
        if (path.size() >= 2) { DrawingGeometry dg; dg.path = path; geoms.push_back(dg); }
    }
    return geoms;
}
